API with and without null-terminated strings

MakisH opened the pull request:
Closes #1, implementing (to the extent I understand) the suggestions of @ivan-pi.

I also tried to make the length an optional argument, to maintain compatibility with user code. I am also trying to avoid touching the preCICE C/Fortran bindings for now, as we just had a breaking release.

Currently, building the (unmodified) `solverdummy.f90` leads to the following error:

```
fortran-module/examples/solverdummy [nullterminated]$ make
gfortran -g solverdummy.f90 -o solverdummy -I../.. -L/home/gc/repos/precice/precice/build -lprecice
/usr/bin/ld: /tmp/ccQ2PSRq.o: in function `MAIN__':
/home/gc/repos/precice/fortran-module/examples/solverdummy/solverdummy.f90:38: undefined reference to `__precice_MOD_precicef_create'
```

Any clue what I am doing wrong?

MakisH committed:
WIP: Add a wrapper precicef_create and a  precicec_create expecting null-terminated strings

ivan-pi commented on precice.f90:
```diff
@@ -3,8 +3,13 @@ module precice
+    !   the corresponding `precicec_<name>`.
   
-    subroutine precicef_create(participantName, configFileName, &
+    subroutine precicec_create(participantName, configFileName, &
       &                        solverProcessIndex, solverProcessSize, &
       &                        participantNameLength, configFileNameLength) &
       &  bind(c, name='precicef_create_')
```
```suggestion
    subroutine precicec_create(participantName, configFileName, &
      &                        solverProcessIndex, solverProcessSize) &
      &  bind(c, name='precicec_createParticipant')
```
If you want to use the C binding and null-terminated strings, it should match the C API:
https://github.com/precice/precice/blob/main/extras/bindings/c/src/preciceC.cpp#L43-L47

MakisH replied:
Ah, I see. So the trick is to target directly the C bindings. Now I get it, thanks!

ivan-pi commented on precice.f90:
```diff
@@ -16,7 +21,7 @@ subroutine precicef_create(participantName, configFileName, &
       integer(kind=c_int) :: solverProcessSize
       integer(kind=c_int), value :: participantNameLength
       integer(kind=c_int), value :: configFileNameLength
```
```suggestion
```

The lengths aren't part of the C API.

ivan-pi commented on precice.f90:
```diff
@@ -312,4 +317,32 @@ end subroutine precicef_get_version_information
+    character(len=*), intent(in) :: configFileName
+    integer(c_int), intent(in) :: solverProcessIndex
+    integer(c_int), intent(in) :: solverProcessSize
+    integer(kind=c_int), intent(in), optional :: participantNameLengthIn
+    integer(kind=c_int), intent(in), optional :: configFileNameLengthIn
+    integer(kind=c_int) participantNameLength
+    integer(kind=c_int) configFileNameLength
+
+    participantNameLength = 50
+    configFileNameLength = 50
+    if(present(participantNameLengthIn)) participantNameLength = participantNameLengthIn
+    if(present(configFileNameLengthIn)) configFileNameLength = configFileNameLengthIn
```
```suggestion
```
Not needed when wrapping the C API.

ivan-pi commented on precice.f90:
```diff
@@ -312,4 +317,32 @@ end subroutine precicef_get_version_information
+    call precicec_create( &
+      trim(participantName)//c_null_char, &
+      trim(configFileName)//c_null_char, &
+      solverProcessIndex, solverProcessSize, &
+      participantNameLength, configFileNameLength)
```
```suggestion
      solverProcessIndex, solverProcessSize)
```

MakisH committed:
Apply suggestions from code review

Co-authored-by: Ivan Pribec <ivan.pribec@gmail.com>

MakisH committed:
Remove length arguments from precicef_create

MakisH commented on precice.f90:
```diff
@@ -312,4 +314,21 @@ end subroutine precicef_get_version_information
 
   end interface
 
+  contains
```
I still want that users call `precicef_create`, so they should be outside the `contains` block. However, it looks like I cannot call subroutines from inside an interface. What would be the right way here?

ivan-pi replied:
But you have it now as a module procedure (a thin wrapper of the preCICE C API). Isn't that what you want?

MakisH replied:
I want that users call the `precicef_create` variant without having to provide neither null-terminated strings, nor placeholder lengths. For this, I would need the `precicef_create` to not be internal (inside the `contains`), no?

If I ask the users to call the `precicec_create`, then they need to provide null-terminated strings.

Should I maybe split it into two different modules?

ivan-pi replied:
In Fortran parlance, interface blocks are like function declarations in C. They are _specification_ statements (i.e. non-executable). 

When you want to _define_ a procedure, it must be in the "contains" part:
```
module precice
 ! ... specification ...
contains
 ! ... procedure definitions (including executable statements) ...
end module
```

MakisH replied:
Ah, that helps! But then, I understand that both what I have in the interface and in the `contains` need to have the same name.

MakisH committed:
Remove the sizes from solverdummy.f90

MakisH committed:
Add .gitignore in example

MakisH committed:
Switch to the Fortran calling convention

MakisH committed:
Improve style in solverdummy example

MakisH committed:
Merge branch 'develop' into nullterminated

MakisH commented:
Together with @ivan-pi, we took now a different approach:

Instead of differentiating between null- and non-null-terminated strings in the C interface, we just adopt the de facto Fortran calling convention, as already done in the core library.

We dropped the `bind(c)` attribute. The way we now write it, automatically appends a trailing `_` in each function name in the binary.

We also removed the dimension arguments present in the `bind(c)` version, and replaced them with `len=*` arguments. This enables the hidden length argument added by the compiler.

Important here would be that this would need special handling on Windows Fortran compilers, which adopt a different convention. We would need to specify [`FortranCInterface`](https://cmake.org/cmake/help/latest/module/FortranCInterface.html#module:FortranCInterface) in the core library `CMakeLists.txt`, to automatically detect the right convention (add an underscore or not). This would assume the presence of a Fortran compiler.

We modified the `precicef_create` and improved the style of the example, I will adapt the rest of the files and ask for a review.

MakisH committed:
Remove ERROR STOP: needs Fortran 2008

MakisH committed:
Remove standard specification from Makefile (unnecessary)
